src/pem.rs

```diff
@@ -3,65 +3,178 @@ use alloc::format;
+    fn from_pem(kind: SectionKind, der: Vec<u8>) -> Option<Self> {
+        match Self::KIND == kind {
+            true => Some(Self::from(der)),
+            false => None,
+        }
+    }
+}
+
+/// Extract and return all PEM sections by reading `rd`.
+#[cfg(feature = "std")]
+pub struct ReadIter<R, T> {
+    rd: R,
+    _ty: PhantomData<T>,
+}
+
+#[cfg(feature = "std")]
+impl<R: io::BufRead, T: PemObject> ReadIter<R, T> {
+    /// Create a new iterator.
+    pub fn new(rd: R) -> Self {
+        Self {
+            rd,
+            _ty: PhantomData,
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl<R: io::BufRead, T: PemObject> Iterator for ReadIter<R, T> {
+    type Item = Result<T, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            return match from_buf(&mut self.rd) {
+                Ok(Some((sec, item))) => match T::from_pem(sec, item) {
+                    Some(res) => Some(Ok(res)),
+                    None => continue,
+                },
+                Ok(None) => return None,
+                Err(err) => Some(Err(err)),
+            };
+        }
+    }
+}
+
+/// Iterator over all PEM sections in a `&[u8]` slice.
+pub struct SliceIter<'a, T> {
+    current: &'a [u8],
+    _ty: PhantomData<T>,
+}
+
+impl<'a, T: PemObject> SliceIter<'a, T> {
+    /// Create a new iterator.
+    pub fn new(current: &'a [u8]) -> Self {
+        Self {
+            current,
+            _ty: PhantomData,
+        }
+    }
+}
+
+impl<T: PemObject> Iterator for SliceIter<'_, T> {
+    type Item = Result<T, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            return match from_slice(self.current) {
+                Ok(Some(((sec, item), rest))) => {
+                    self.current = rest;
+                    match T::from_pem(sec, item) {
+                        Some(res) => Some(Ok(res)),
+                        None => continue,
+                    }
+                }
+                Ok(None) => return None,
+                Err(err) => Some(Err(err)),
+            };
+        }
+    }
+}
+
+impl PemObject for (SectionKind, Vec<u8>) {
+    fn from_pem(kind: SectionKind, der: Vec<u8>) -> Option<Self> {
+        Some((kind, der))
+    }
+}
+
+/// Extract and decode the next supported PEM section from `input`
 ///
 /// - `Ok(None)` is returned if there is no PEM section to read from `input`
 /// - Syntax errors and decoding errors produce a `Err(...)`
-/// - Otherwise each decoded section is returned with a `Ok(Some((Item::..., remainder)))` where
+/// - Otherwise each decoded section is returned with a `Ok(Some((..., remainder)))` where
 ///   `remainder` is the part of the `input` that follows the returned section
-pub fn read_one_from_slice(mut input: &[u8]) -> Result<Option<(Item, &[u8])>, Error> {
+#[allow(clippy::type_complexity)]
+pub fn from_slice(mut input: &[u8]) -> Result<Option<((SectionKind, Vec<u8>), &[u8])>, Error> {
```

Comment:
I'm inclined to think that we should keep this private for now? We have a public `SliceIter::new()`.

Reply:
I think it would be good to keep this private; I would need to add a `pub fn remainder(&self) -> &[u8] { ... }` (ala the chunk iterators in core) to `SliceIter`

Reply:
Makes sense. Maybe make that `doc(hidden)`?